pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,459 @@
+		info, err := peer.AddrInfoFromP2pAddr(addr)
+		if err != nil {
+			t.logger.Log(logging.LevelWarn, "connect: failed to parse addr", "src", t.ownID, "dest", nodeID, "addr", addr, "err", err)
+			continue
+		}
+		parsedAddrInfo[nodeID] = info
+	}
+
+	t.connsLock.Lock()
+	for nodeID := range parsedAddrInfo {
+		if nodeID == t.ownID {
+			continue
+		}
+		_, found := t.conns[nodeID]
+		if !found {
+			t.conns[nodeID] = &connInfo{AddrInfo: parsedAddrInfo[nodeID], Stream: nil}
+		}
+	}
+	t.connsLock.Unlock()
+
+	t.connect(ctx, nodes)
+}
+
+func (t *Transport) Send(ctx context.Context, dest types.NodeID, msg *messagepb.Message) error {
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	out, err := t.encode(msg)
+	if err != nil {
+		return err
+	}
+
+	err = t.sendPayload(dest, out)
+	if err != nil {
+		t.reconnect(ctx, dest)
+		return errors.Wrapf(err, "%s failed to send data to %s", t.ownID, dest)
+	}
+
+	return nil
+}
+
+func (t *Transport) ImplementsModule() {}
+
+func (t *Transport) EventsOut() <-chan *events.EventList {
+	return t.incomingMessages
+}
+
+func (t *Transport) ApplyEvents(ctx context.Context, eventList *events.EventList) error {
+	iter := eventList.Iterator()
+	for event := iter.Next(); event != nil; event = iter.Next() {
+
+		switch e := event.Type.(type) {
+		case *eventpb.Event_Init:
+			// no actions on init
+		case *eventpb.Event_SendMessage:
+			for _, destID := range e.SendMessage.Destinations {
+				if types.NodeID(destID) == t.ownID {
+					// Send message to myself bypassing the network.
+					// The sending must be done in its own goroutine in case writing to gt.incomingMessages blocks.
+					// (Processing of input events must be non-blocking.)
+					receivedEvent := events.MessageReceived(
+						types.ModuleID(e.SendMessage.Msg.DestModule),
+						t.ownID,
+						e.SendMessage.Msg,
+					)
+					go func() {
+						select {
+						case t.incomingMessages <- events.ListOf(receivedEvent):
+						case <-ctx.Done():
+						}
+					}()
+				} else {
+					// Send message to another node.
+					if err := t.Send(ctx, types.NodeID(destID), e.SendMessage.Msg); err != nil {
+						t.logger.Log(logging.LevelError, "failed to send a message", "err", err)
+					}
+				}
+			}
+		default:
+			return fmt.Errorf("unexpected event: %T", event.Type)
+		}
+	}
+
+	return nil
+}
+
+func (t *Transport) connect(ctx context.Context, nodes map[types.NodeID]types.NodeAddress) {
+	t.connWg.Add(1)
+
+	go func() {
+		defer t.connWg.Done()
+
+		t.connsLock.Lock()
+		defer t.connsLock.Unlock()
+
+		t.logger.Log(logging.LevelDebug, "started connecting nodes", "src", t.ownID)
+		defer t.logger.Log(logging.LevelDebug, "finished connecting nodes", "src", t.ownID)
+
+		wg := &sync.WaitGroup{}
+		wg.Add(len(nodes))
+
+		for nodeID := range nodes {
+			// Do not establish a real connection with own node.
+			if nodeID == t.ownID {
+				wg.Done()
+				continue
+			}
+
+			go t.connectToNodeWithoutLock(ctx, nodeID, wg)
```

Comment:
If I remember correctly the main requirement is that connection and sending must be nonblocking. Correct?
If correct then we need a goroutine over connecting goroutines, otherwise, this call will be blocking.
If you meant by "nonblocking" something else then define it and I will reimplement it.